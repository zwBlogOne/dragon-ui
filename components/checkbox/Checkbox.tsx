import React, { Component } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';

class Checkbox extends Component<PropsType, any> {
  static Group;

  static defaultProps = {
    prefixCls: 'ui-checkbox',
    defaultChecked: false,
    isDisabled: false,
    indeterminate: false,
    onChange: () => {},
  };

  constructor(props) {
    super(props);
    this.state = {
      checked: props.checked || props.defaultChecked,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: !!nextProps.checked,
      });
    }
  }

  _onClick(e) {
    const { checked } = this.state;

    this.setState({
      checked: !checked,
    });
    this.props.onChange(e);
  }

  render() {
    const { props } = this;
    const {
      prefixCls,
      value,
      isDisabled,
      className,
      children,
      style,
      indeterminate,
    } = props;
    const disabled = 'disabled' in props || isDisabled;
    const cls = classnames({
      [prefixCls!]: true,
      checked: this.state.checked,
      disabled,
      indeterminate: this.state.checked && indeterminate,
      [className!]: !!className,
    });

    return (
      <label style={style}>
        <span className={cls}>
          <input
            className={`${prefixCls}-input`}
            type="checkbox"
            value={value}
            checked={this.state.checked}
            disabled={disabled}
            onChange={e => this._onClick(e)}
          />
          <span className={`${prefixCls}-inner`} />
        </span>
        {children}
      </label>
    );
  }
}

export default Checkbox;
