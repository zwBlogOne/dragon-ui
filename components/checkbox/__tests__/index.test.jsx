import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Checkbox from '../index';

describe('Checkbox', () => {
  it('renders normal Checkbox correctly', () => {
    const wrapper = render(
      <Checkbox
        checked
        value="0"
      >
        选择0
      </Checkbox>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders disabled Checkbox correctly', () => {
    const wrapper = render(
      <Checkbox
        checked
        disabled
        value="0"
      >
        选择0
      </Checkbox>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders indeterminate Checkbox correctly', () => {
    const wrapper = render(
      <Checkbox
        indeterminate
        value="0"
      >
        选择0
      </Checkbox>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders disabled CheckboxGroup correctly', () => {
    const wrapper = render(
      <Checkbox.Group>
        <Checkbox value="a">A</Checkbox>
        <Checkbox value="b">B</Checkbox>
        <Checkbox value="c">C</Checkbox>
        <Checkbox value="d">D</Checkbox>
      </Checkbox.Group>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders CheckboxGroup correctly with defaultValue', () => {
    const wrapper = render(
      <Checkbox.Group defaultValue={['a']}>
        <Checkbox value="a">A</Checkbox>
        <Checkbox value="b">B</Checkbox>
        <Checkbox value="c">C</Checkbox>
        <Checkbox value="d">D</Checkbox>
      </Checkbox.Group>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when clicking on Checkbox', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Checkbox
        onChange={onChange}
        value="0"
      >
        选择0
      </Checkbox>
    );

    wrapper.find('input').simulate('change');

    expect(onChange).toHaveBeenCalled();
  });

  it('behaves correctly when clicking on Checkbox', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Checkbox
        onChange={onChange}
        value="0"
      >
        选择0
      </Checkbox>
    );

    wrapper.find('input').simulate('change');

    expect(onChange).toHaveBeenCalled();
  });

  it('behaves correctly when checking in CheckboxGroup', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Checkbox.Group onChange={onChange}>
        <Checkbox value="a">A</Checkbox>
        <Checkbox value="b">B</Checkbox>
        <Checkbox value="c">C</Checkbox>
        <Checkbox value="d">D</Checkbox>
      </Checkbox.Group>
    );

    wrapper.find('input').first().simulate('change');

    expect(onChange).toHaveBeenCalledWith(['a']);
  });

  it('behaves correctly when CheckboxGroup receives new value', () => {
    const wrapper = mount(
      <Checkbox.Group value={[]}>
        <Checkbox value="a">A</Checkbox>
        <Checkbox value="b">B</Checkbox>
        <Checkbox value="c">C</Checkbox>
        <Checkbox value="d">D</Checkbox>
      </Checkbox.Group>
    );

    expect(wrapper.find('.ui-checkbox').first().hasClass('checked')).toBeFalsy();
    wrapper.setProps({ value: ['a'] });
    expect(wrapper.find('.ui-checkbox').first().hasClass('checked')).toBeTruthy();
  });
});
