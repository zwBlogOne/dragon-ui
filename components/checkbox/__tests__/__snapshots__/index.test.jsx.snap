// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders CheckboxGroup correctly with defaultValue 1`] = `
<div
  class="ui-checkbox-group"
>
  <label>
    <span
      class="ui-checkbox checked"
    >
      <input
        checked=""
        class="ui-checkbox-input"
        type="checkbox"
        value="a"
      />
      <span
        class="ui-checkbox-inner"
      />
    </span>
    A
  </label>
  <label>
    <span
      class="ui-checkbox"
    >
      <input
        class="ui-checkbox-input"
        type="checkbox"
        value="b"
      />
      <span
        class="ui-checkbox-inner"
      />
    </span>
    B
  </label>
  <label>
    <span
      class="ui-checkbox"
    >
      <input
        class="ui-checkbox-input"
        type="checkbox"
        value="c"
      />
      <span
        class="ui-checkbox-inner"
      />
    </span>
    C
  </label>
  <label>
    <span
      class="ui-checkbox"
    >
      <input
        class="ui-checkbox-input"
        type="checkbox"
        value="d"
      />
      <span
        class="ui-checkbox-inner"
      />
    </span>
    D
  </label>
</div>
`;

exports[`Checkbox renders disabled Checkbox correctly 1`] = `
<label>
  <span
    class="ui-checkbox checked disabled"
  >
    <input
      checked=""
      class="ui-checkbox-input"
      disabled=""
      type="checkbox"
      value="0"
    />
    <span
      class="ui-checkbox-inner"
    />
  </span>
  选择0
</label>
`;

exports[`Checkbox renders disabled CheckboxGroup correctly 1`] = `
<div
  class="ui-checkbox-group"
>
  <label>
    <span
      class="ui-checkbox"
    >
      <input
        class="ui-checkbox-input"
        type="checkbox"
        value="a"
      />
      <span
        class="ui-checkbox-inner"
      />
    </span>
    A
  </label>
  <label>
    <span
      class="ui-checkbox"
    >
      <input
        class="ui-checkbox-input"
        type="checkbox"
        value="b"
      />
      <span
        class="ui-checkbox-inner"
      />
    </span>
    B
  </label>
  <label>
    <span
      class="ui-checkbox"
    >
      <input
        class="ui-checkbox-input"
        type="checkbox"
        value="c"
      />
      <span
        class="ui-checkbox-inner"
      />
    </span>
    C
  </label>
  <label>
    <span
      class="ui-checkbox"
    >
      <input
        class="ui-checkbox-input"
        type="checkbox"
        value="d"
      />
      <span
        class="ui-checkbox-inner"
      />
    </span>
    D
  </label>
</div>
`;

exports[`Checkbox renders indeterminate Checkbox correctly 1`] = `
<label>
  <span
    class="ui-checkbox"
  >
    <input
      class="ui-checkbox-input"
      type="checkbox"
      value="0"
    />
    <span
      class="ui-checkbox-inner"
    />
  </span>
  选择0
</label>
`;

exports[`Checkbox renders normal Checkbox correctly 1`] = `
<label>
  <span
    class="ui-checkbox checked"
  >
    <input
      checked=""
      class="ui-checkbox-input"
      type="checkbox"
      value="0"
    />
    <span
      class="ui-checkbox-inner"
    />
  </span>
  选择0
</label>
`;
